src/util/result.h

```diff
@@ -31,54 +140,110 @@ struct Error {
 //! `std::optional<T>` can be updated to return `util::Result<T>` and return
 //! error strings usually just replacing `return std::nullopt;` with `return
 //! util::Error{error_string};`.
-template <class T>
-class Result
+//!
+//! Most code does not need different error-handling behavior for different
+//! types of errors, and can suffice just using the type `T` success value on
+//! success, and descriptive error strings when there's a failure. But
+//! applications that do need more complicated error-handling behavior can
+//! override the default `F = void` failure type and get failure values by
```

Comment:
Allowing for a `void` failure type seems to make this incompatible to be switched out to `std::expected` https://en.cppreference.com/w/cpp/utility/expected ?

With multiple warning and error messages this may already be incompatible, though?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r1104282294

> Allowing for a `void` failure type seems to make this incompatible to be switched out to `std::expected` https://en.cppreference.com/w/cpp/utility/expected ?
> 
> With multiple warning and error messages this may already be incompatible, though?

Yes, if you are thinking that `util::Result` could wrap `std::expected` that would be probably be awkward and not worth it.

But I don't think there is a conflict because the two classes are mostly doing different things. The `util::Result` class is mostly providing error-reporting functionality (passing error and warning strings up to the user). The `std::expected` class is only providing error-handling functionality (passing failure and success values between functions). Here is how I would choose the between the two classes:

- If your function never fails, it should just return success value directly.
- If your function can fail, but doesn't provide any error strings or specific failure information, it should return `std::optional`.
- If your function can fail, and provides failure information but not error strings, it should return `std::expected`.
- If your function can fail, and generates error or warning strings it should return `util::Result`.

We have a lot of functions that generate error strings as you can see by all the code using `util::Error` and `util::Result` presently, and in more code that is converted to use `util::Result` in this PR and #25722. After `std::expected` is available, most of these functions still be better off using `util::Result` instead of `std::expected` so they are able to pass back error strings in a uniform way.

But when `std::expected` is available, we may want to tweak the `util::Result` class to make it easier to switch between `std::expected` and `util::Result` with minimal code changes. For example, the `util::Result` already has a `value_or` method to be compatible with `std::optional`. It could also have `and_then` and `or_else` methods to be compatible with `std::expected`.